internal/oci/notation.go

```diff
@@ -0,0 +1,225 @@
+// remote options used by the verifier
+func WithNotaryRemoteOptions(opts ...remote.Option) NotationOptions {
+	return func(o *notationOptions) {
+		o.ROpt = opts
+	}
+}
+
+// WithNotaryKeychain is a functional option for overriding the default
+// remote options used by the verifier
+func WithNotaryKeychain(key authn.Keychain) NotationOptions {
+	return func(o *notationOptions) {
+		o.Keychain = key
+	}
+}
+
+// NotatryVerifier is a struct which is responsible for executing verification logic
+type NotaryVerifier struct {
+	auth     authn.Keychain
+	verifier *notation.Verifier
+	opts     []remote.Option
+	insecure bool
+}
+
+// NewNotaryVerifier initializes a new NotaryVerifier
+func NewNotaryVerifier(opts ...NotationOptions) (*NotaryVerifier, error) {
+	o := notationOptions{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	for _, pol := range o.TrustStore.TrustPolicies {
+		certName := pol.Name
+		for _, store := range pol.TrustStores {
+			s := strings.Split(store, ":")
+			if len(s) != 2 {
+				return nil, fmt.Errorf("trust store '%s' is invalid. Trust store name must contain a store type and a store name separated by ':'. For example 'ca:fluxcd.io'", store)
+			}
+			generateTrustStore(s[0], s[1], certName, o.PublicKey)
+		}
+	}
+
+	verifier, err := verifier.New(o.TrustStore, truststore.NewX509TrustStore(dir.ConfigFS()), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &NotaryVerifier{
+		auth:     o.Keychain,
+		verifier: &verifier,
+		opts:     o.ROpt,
+		insecure: o.Insecure,
+	}, nil
+}
+
+// Verify verifies the authenticity of the given ref OCI image.
+// It returns a boolean indicating if the verification was successful.
+// It returns an error if the verification fails, nil otherwise.
+func (v *NotaryVerifier) Verify(ctx context.Context, ref name.Reference) (bool, error) {
+	url := ref.Name()
+	remoteRepo, err := oras.NewRepository(url)
+	if err != nil {
+		return false, err
+	}
+	remoteRepo.PlainHTTP = v.insecure
+
+	repo := registry.NewRepository(remoteRepo)
+
+	ss := stringResource{url}
+
+	var credentialProvider func(ctx context.Context, registry string) (oauth.Credential, error)
+
+	if v.auth != nil {
+		au, err := v.auth.Resolve(ss)
+		if err != nil {
+			return false, err
+		}
+
+		authConfig, err := au.Authorization()
+		if err != nil {
+			return false, err
+		}
+
+		credentialProvider = func(ctx context.Context, registry string) (oauth.Credential, error) {
+			if authConfig.Username != "" || authConfig.Password != "" || authConfig.IdentityToken != "" {
+				return oauth.Credential{
+					Username:     authConfig.Username,
+					Password:     authConfig.Password,
+					RefreshToken: authConfig.IdentityToken,
+				}, nil
+			}
+			return oauth.EmptyCredential, nil
+		}
+	}
+
+	repoClient := &oauth.Client{
+		Header: http.Header{
+			"User-Agent": {"flux"},
+		},
+		Cache:      oauth.NewCache(),
+		Credential: credentialProvider,
+	}
+
+	remoteRepo.Client = repoClient
+
+	i, err := remote.Image(ref, v.opts...)
```

Comment:
I wish we could just use `github.com/google/go-containerregistry` and don't mix this with `oras/v2`. This would make this more coherent as reduce our dependencies. It seems that `notary` force us to dependent on `oras/v2` here: `repo := registry.NewRepository(remoteRepo)`.

Reply:
Tell me about it! If notary had used `github.com/google/go-containerregistry` my life would have been so much easier 😅